### PR TITLE
CB-17714 Support in Persister to allocate the Resources in batch inst…

### PR DESCRIPTION
**Pull request by topolyai5**
…ead of one by one

See detailed description in the commit message.

### CODE REVIEW

File: cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsLaunchTest.java [hunk 1]
```diff
@@ -215,8 +215,9 @@ public void launchStack() throws Exception {
 
         // assert
         verify(persistenceNotifier).notifyAllocation(argThat(cloudResource -> ResourceType.AWS_VPC.equals(cloudResource.getType())), any());
-        verify(persistenceNotifier, times(2))
-                .notifyAllocation(argThat(cloudResource -> ResourceType.AWS_VOLUMESET.equals(cloudResource.getType())), any());
+        verify(persistenceNotifier, times(2)).notifyAllocations(
```
Comment (bergerdenes) [inline cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsLaunchTest.java:218]: should be `times(1)`?

File: cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsLaunchTest.java [hunk 2]
```diff
@@ -260,8 +261,9 @@ public void launchStackWithEfs() throws Exception {
...
```
Comment (bergerdenes) [inline cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsLaunchTest.java:264]: should be `times(1)`?

File: cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsRepairTest.java [hunk 3]
```diff
@@ -371,15 +372,16 @@ private void upscaleStack() throws Exception {
                 .suspendProcesses(argThat(argument -> AUTOSCALING_GROUP_NAME.equals(argument.getAutoScalingGroupName())
                         && SUSPENDED_PROCESSES.equals(argument.getScalingProcesses())));
 
-        ArgumentCaptor<CloudResource> updatedCloudResourceArgumentCaptor = ArgumentCaptor.forClass(CloudResource.class);
-        verify(resourceNotifier, times(2)).notifyUpdate(updatedCloudResourceArgumentCaptor.capture(), any());
+        ArgumentCaptor<List<CloudResource>> updatedCloudResourceArgumentCaptor = ArgumentCaptor.forClass(List.class);
+        verify(resourceNotifier, times(2)).notifyUpdates(updatedCloudResourceArgumentCaptor.capture(), any());
```
Comment (bergerdenes) [inline cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsRepairTest.java:376]: should be `times(1)`?

File: cloud-reactor-api/src/main/java/com/sequenceiq/cloudbreak/cloud/event/resource/LaunchLoadBalancerResult.java [hunk 4]
```diff
@@ -5,7 +5,7 @@
 import com.sequenceiq.cloudbreak.cloud.event.CloudPlatformResult;
 import com.sequenceiq.cloudbreak.cloud.model.CloudResourceStatus;
 
-public class LaunchLoadBalancerResult extends CloudPlatformResult {
+public class  LaunchLoadBalancerResult extends CloudPlatformResult {
```
Comment (bergerdenes) [inline cloud-reactor-api/src/main/java/com/sequenceiq/cloudbreak/cloud/event/resource/LaunchLoadBalancerResult.java:8]: extra space

File: core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/CloudResourcePersisterService.java [hunk 5]
```diff
@@ -37,40 +40,53 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
+                .filter(cr -> !resourceExists(stackId, cr))
+                .map(cloudResource -> {
+                    Resource resource = cloudResourceToResourceConverter.convert(cloudResource);
+                    setStack(stack, cloudResource, resource);
+                    resourceService.save(resource);
+                    return resource;
+                })
+                .collect(Collectors.toList());
+        if (cloudResources.size() != resources.size()) {
+            LOGGER.debug("There are {} resource(s) which won't save", cloudResources.size() - resources.size());
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/CloudResourcePersisterService.java:55]: won't be saved

File: core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/CloudResourcePersisterService.java [hunk 6]
```diff
@@ -37,40 +40,53 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
-        persistedResource.ifPresent(value -> resourceService.delete(value));
+        List<CloudResource> cloudResources = notification.getCloudResources();
+        AtomicInteger deleted = new AtomicInteger(0);
+        cloudResources.stream()
+                .filter(cr -> resourceExists(stackId, cr))
+                .forEach(cloudResource -> {
+                    deleteResource(stackId, cloudResource);
+                    deleted.incrementAndGet();
+                });
+        LOGGER.debug("There are deleted {} resource(s)", deleted.get());
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/CloudResourcePersisterService.java:89]: There are {} deleted resource(s)

File: core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/CloudResourcePersisterService.java [hunk 7]
```diff
@@ -93,10 +109,34 @@ private Optional<Resource> getPersistedResource(Long stackId, CloudResource clou
...
-            LOGGER.debug("Setting stack {} for cloud resource {} and type {}", stackId, cloudResource.getName(), cloudResource.getType());
-            resource.setStack(findStackById(stackId));
+            exists = resourceService.existsByStackIdAndNameAndType(stackId, cloudResource.getName(), cloudResource.getType());
+            id = cloudResource.getName();
+        } else {
+            exists = resourceService.existsByResourceReferenceAndType(cloudResource.getReference(), cloudResource.getType());
+            id = cloudResource.getReference();
+        }
+        if (exists) {
+            LOGGER.debug("{} and {} is already exists in DB for stack.", id, cloudResource.getType());
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/eventbus/CloudResourcePersisterService.java:123]: ~~is~~

File: environment/src/main/java/com/sequenceiq/environment/resourcepersister/CloudResourcePersisterService.java [hunk 8]
```diff
@@ -33,40 +35,59 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
-                ResourceType.valueOf(cloudResource.getType().name())).ifPresent(value -> resourceService.delete(value));
+        List<CloudResource> cloudResources = notification.getCloudResources();
+        AtomicInteger deleted = new AtomicInteger(0);
+        cloudResources.stream()
+                .filter(cr -> resourceService.existsByResourceReferenceAndType(cr.getReference(), cr.getType()))
+                .forEach(cloudResource -> {
+                    resourceService.deleteByReferenceAndType(cloudResource.getReference(), cloudResource.getType());
+                    deleted.incrementAndGet();
+                });
+        LOGGER.debug("There are deleted {} resource(s)", deleted.get());
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/resourcepersister/CloudResourcePersisterService.java:90]: There are {} deleted resource(s)

File: environment/src/main/java/com/sequenceiq/environment/resourcepersister/ResourceRepository.java [hunk 9]
```diff
@@ -28,9 +29,15 @@ Optional<Resource> findByResourceReferenceAndStatusAndType(@Param("resourceRefer
...
+    Optional<Resource> findByResourceReferenceAndType(@Param("resourceReference") String resourceReference, @Param("type") ResourceType type);
 
     @Query("SELECT r FROM Resource r WHERE r.environment.id = :environmentId AND r.resourceType = :type")
     Optional<Resource> findByEnvironmentIdAndType(@Param("environmentId") Long environmentId, @Param("type") com.sequenceiq.common.api.type.ResourceType type);
+
+    @Query("SELECT count(r) > 0 FROM Resource r WHERE r.resourceReference = :resourceReference AND r.resourceType = :type")
+    boolean existsByResourceReferenceAndType(@Param("resourceReference") String resourceReference, @Param("type") ResourceType type);
+
+    @Modifying
+    @Query("DELETE FROM Resource r WHERE  r.resourceReference = :resourceReference AND r.resourceType = :type")
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/resourcepersister/ResourceRepository.java:41]: extra space

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/cloud/CloudResourcePersisterService.java [hunk 10]
```diff
@@ -37,40 +39,53 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
+                .filter(cr -> !resourceExists(stackId, cr))
+                .map(cloudResource -> {
+                    Resource resource = cloudResourceToResourceConverter.convert(cloudResource);
+                    setStack(stack, cloudResource, resource);
+                    resourceService.save(resource);
+                    return resource;
+                })
+                .collect(Collectors.toList());
+        if (cloudResources.size() != resources.size()) {
+            LOGGER.debug("There are {} resource(s), these will not save", cloudResources.size() - resources.size());
```
Comment (bergerdenes) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/cloud/CloudResourcePersisterService.java:54]: will not be saved

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/cloud/CloudResourcePersisterService.java [hunk 11]
```diff
@@ -37,40 +39,53 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
-        persistedResource.ifPresent(value -> resourceService.delete(value));
+        List<CloudResource> cloudResources = notification.getCloudResources();
+        AtomicInteger deleted = new AtomicInteger(0);
+        cloudResources.stream()
+                .filter(cr -> resourceExists(stackId, cr))
+                .forEach(cloudResource -> {
+                    resourceService.deleteByStackIdAndNameAndType(stackId, cloudResource.getName(), cloudResource.getType());
+                    deleted.incrementAndGet();
+                });
+        LOGGER.debug("There are deleted {} resource(s)", deleted.get());
```
Comment (bergerdenes) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/cloud/CloudResourcePersisterService.java:88]: There are {} deleted resource(s)

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/cloud/CloudResourcePersisterService.java [hunk 12]
```diff
@@ -93,11 +108,26 @@ private Optional<Resource> getPersistedResource(Long stackId, CloudResource clou
...
+        String id;
+        if (cloudResource.isStackAware()) {
+            exists = resourceService.existsByStackIdAndNameAndType(stackId, cloudResource.getName(), cloudResource.getType());
+            id = cloudResource.getName();
+        } else {
+            exists = resourceService.existsByResourceReferenceAndType(cloudResource.getReference(), cloudResource.getType());
+            id = cloudResource.getReference();
+        }
+        if (exists) {
+            LOGGER.debug("{} and {} is already exists in DB for stack.", id, cloudResource.getType());
```
Comment (bergerdenes) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/cloud/CloudResourcePersisterService.java:129]: ~~is~~

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/cloud/CloudResourcePersisterService.java [hunk 13]
```diff
@@ -36,30 +38,43 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
-            LOGGER.debug("Trying to persist a resource (name: {}, type: {}, stackId: {}) that is already persisted, skipping..",
-                    cloudResource.getName(), cloudResource.getType().name(), dbStackId);
-            return notification;
+        DBStack dbStack = dbStackService.getById(dbStackId);
+        List<CloudResource> cloudResources = notification.getCloudResources();
+        List<DBResource> resources = cloudResources.stream()
+                .filter(cr -> {
+                    boolean exist = dbResourceService.existsByStackAndNameAndType(dbStackId, cr.getName(), cr.getType());
+                    if (exist) {
+                        LOGGER.debug("{} and {} is already exists in DB for stack.", cr.getName(), cr.getType());
```
Comment (bergerdenes) [inline redbeams/src/main/java/com/sequenceiq/redbeams/service/cloud/CloudResourcePersisterService.java:47]: ~~is~~

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/cloud/CloudResourcePersisterService.java [hunk 14]
```diff
@@ -36,30 +38,43 @@ public class CloudResourcePersisterService implements Persister<ResourceNotifica
...
+                })
+                .map(cloudResource -> {
+                    DBResource resource = cloudResourceToDbResourceConverter.convert(cloudResource);
+                    resource.setDbStack(dbStack);
+                    dbResourceService.save(resource);
+                    return resource;
+                })
+                .collect(Collectors.toList());
+        if (cloudResources.size() != resources.size()) {
+            LOGGER.debug("There are {} resource(s), these will not save", cloudResources.size() - resources.size());
```
Comment (bergerdenes) [inline redbeams/src/main/java/com/sequenceiq/redbeams/service/cloud/CloudResourcePersisterService.java:59]: will not be saved

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsLaunchTest.java (hunk 1 above)
Comment (topolyai5) [reply cloud-aws-cloudformation/src/test/java/com/sequenceiq/cloudbreak/cloud/aws/component/AwsLaunchTest.java:218]: nope, it is green

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**